app/grandchallenge/notifications/models.py

```diff
@@ -1,27 +1,119 @@
-from actstream.models import Action
+from actstream.models import Action, followers
 from django.conf import settings
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.contrib.humanize.templatetags.humanize import naturaltime
 from django.db import models
+from django.utils.html import format_html
+from django.utils.translation import gettext_lazy as _
 from guardian.shortcuts import assign_perm
 
 from grandchallenge.core.models import UUIDModel
+from grandchallenge.profiles.templatetags.profiles import user_profile_link
+from grandchallenge.subdomains.utils import reverse
+
+
+class NotificationTypeChoices(models.TextChoices):
+    """Notification type choices."""
+
+    GENERIC = "GENERIC", _("Generic")
+    FORUM_POST = "FRM-POST", _("Forum post")
+    FORUM_POST_REPLY = "FRM-RPL", _("Forum post reply")
+    ACCESS_REQUEST = "ACC-REQ", _("Access request")
+    REQUEST_UPDATE = "REQ-UPD", _("Request update")
+    NEW_ADMIN = "ADMIN", _("New admin")
+    EVALUATION_STATUS = "EVAL", _("Evaluation status update")
+    MISSING_METHOD = "MISSING-METHOD", _("Missing method")
+    JOB_STATUS = "JOB", _("Job status update")
+    IMAGE_IMPORT_STATUS = "IMAGE-IMPORT", _("Image import status update")
+
+
+class NotificationType:
+    """Notification type."""
+
+    NotificationTypeChoices = NotificationTypeChoices
 
 
 class Notification(UUIDModel):
+    Type = NotificationType.NotificationTypeChoices
 
     user = models.ForeignKey(
         settings.AUTH_USER_MODEL,
         help_text="Which user does this notification correspond to?",
         on_delete=models.CASCADE,
     )
 
+    type = models.CharField(
+        max_length=15,
+        choices=Type.choices,
+        default=Type.GENERIC,
+        help_text=("Of what type is this notification?"),
+    )
+    # TODO: deprecate this field after old notifications have been transferred
     action = models.ForeignKey(
         Action,
+        blank=True,
+        null=True,
         help_text="Which action is associated with this notification?",
         on_delete=models.CASCADE,
     )
 
     read = models.BooleanField(default=False, db_index=True)
 
+    context_class = models.CharField(
+        max_length=10,
+        blank=True,
+        null=True,
+        help_text="Bootstrap contextual class to style notification list items.",
+    )
+
+    # action-related fields (taken from actstream.models.Action)
+    actor_content_type = models.ForeignKey(
+        ContentType,
+        blank=True,
+        null=True,
+        related_name="notification_actor",
+        on_delete=models.CASCADE,
+        db_index=True,
+    )
+    actor_object_id = models.CharField(
+        max_length=255, db_index=True, blank=True, null=True
+    )
+    actor = GenericForeignKey("actor_content_type", "actor_object_id")
+
+    verb = models.CharField(
```

Comment:
I'm being a bit nitpicky here, but `verb` seems like it does not really cover the purpose of this field. If we're recreating the model anyway, I'd prefer to rename this field to something like `message`.

Reply:
Ok, that makes sense actually.